src/circuit/gadgets/constraints.rs

```diff
@@ -698,6 +799,174 @@ mod tests {
+                let pb = Boolean::constant(premise);
+                let _ = implies_equal_zero(&mut cs.namespace(|| "implies equal zero"), &pb, &num);
+                assert_eq!(cs.is_satisfied(), result);
+
+            };
+
+            // any premise
+            test_premise_num(p, 0, true);
+
+            // false premise, any value
+            test_premise_num(false, rand_a, true);
+
+            // true premise, bad values
+            test_premise_num(true, rand_positive, false);
+        }
+
+        #[test]
+        fn test_implies_equal(p in any::<bool>(), (a, b) in any::<(FWrap<Fr>, FWrap<Fr>)>()) {
+            prop_assume!(a != b);
+
+            let test_a_b = |premise: bool, a, b, result: bool| {
+                let mut cs = TestConstraintSystem::<Fr>::new();
+                let a_num = AllocatedNum::alloc(cs.namespace(|| "a_num"), || Ok(Fr::from(a))).unwrap();
+                let b_num = AllocatedNum::alloc(cs.namespace(|| "b_num"), || Ok(Fr::from(b))).unwrap();
+                let pb = Boolean::constant(premise);
+                let _ = implies_equal(&mut cs.namespace(|| "implies equal"), &pb, &a_num, &b_num);
+                assert_eq!(cs.is_satisfied(), result);
+            };
+
+            // any premise
+            test_a_b(p, a.0, a.0, true);
+
+            // positive case
+            test_a_b(false, a.0, b.0, true);
+
+            // negative case
+            test_a_b(true, a.0, b.0, false);
+        }
+
+        #[test]
+        fn test_popcount_equal((i, j, k) in ((0usize..7), (0usize..7), (0usize..7))) {
+            prop_assume!(i != j);
+            prop_assume!(j != k);
+            prop_assume!(i != k);
+
+            let mut v = vec![Boolean::constant(false); 7];
+
+            let mut test_sum = |elem: Option<usize>, sum: u64, result: bool| {
+                let mut cs = TestConstraintSystem::<Fr>::new();
+                if let Some(e) = elem {
+                    v[e] = Boolean::constant(true);
+                };
+                let alloc_sum = AllocatedNum::alloc(cs.namespace(|| "sum"), || Ok(Fr::from(sum)));
+                let _ = popcount_equal(&mut cs.namespace(|| "popcount equal"), &v, alloc_sum.unwrap().get_variable());
+                assert_eq!(cs.is_satisfied(), result);
+            };
+
+            // All values are false, popcount must be zero
+            test_sum(None, 0, true);
+            // Insert true into a random position, now popocount must be one
+            test_sum(Some(i), 1, true);
+            // Insert true into a distinct random position, now popocount must be two
+            test_sum(Some(j), 2, true);
+            // Insert true again into a distinct random position, now popocount must be three
+            test_sum(Some(k), 3, true);
+
+            // negative test, sum can't be a random number between 4 and MAX
+            let rand_wrong = rand::thread_rng().gen_range(4..u64::MAX);
```

Comment:
Can we get this as an input rather than grabbing new randomness within the test?

Reply:
Since we're not sampling the whole range of field elements anyway, would it make more sense to focus on the range that might more plausibly occur in wrong results? I'd be much more concerned that the popcount might be off by one or otherwise be in the range 0..8 than that it be much larger. So maybe 4..12 or something? Just thinking out loud to start a dialog about heuristics we evolve for such ranges.

Reply:
done